timemachine/cpp/src/stream_manager.cu

```diff
@@ -0,0 +1,60 @@
+// this implements a stream manager that allows creation of new streams as well as syncing two streams
+
+#include "gpu_utils.cuh"
+#include "kernel_utils.cuh"
+#include "stream_manager.hpp"
+
+namespace timemachine {
+
+StreamManager::StreamManager(){};
+
+StreamManager::~StreamManager() {
+    for (const auto &[key, value] : streams_) {
+        gpuErrchk(cudaStreamDestroy(value));
+    }
+    for (const auto &[key, value] : events_) {
+        gpuErrchk(cudaEventDestroy(value));
+    }
+}
+
+cudaStream_t StreamManager::get_stream(int key) {
+    if (streams_.count(key) == 1) {
+        return streams_[key];
+    }
+    cudaStream_t new_stream;
+    // Create stream that doesn't block with the null stream
+    gpuErrchk(cudaStreamCreateWithFlags(&new_stream, cudaStreamNonBlocking));
+
+    streams_[key] = new_stream;
+    return new_stream;
+};
+
+cudaEvent_t StreamManager::get_stream_event(int key) {
```

Comment:
Clarification: is the purpose of `get_stream` and `get_stream_event` just to cache streams and events and avoid any overhead of creating new streams and events, or is there a more fundamental reason? Might be worth documenting (naively it seems like we could simplify the code by removing this and making `sync_to` and `sync_from` simple functions)

Reply:
The overhead of creating events and streams are negligible, destruction of streams is not. cudaEventDestroy is async while cudaStreamDestroy is synchronous. Mostly to simply the management of streams and events to avoid having to manage creation and destruction.

Will add some documentation.

Reply:
Added in [9665423](https://github.com/proteneer/timemachine/pull/1024/commits/96654233a86cf5630ed02634c1f13f3ddb0079ce)